fairmq/FairMQDevice.h

```diff
@@ -433,6 +434,16 @@ class FairMQDevice : public FairMQStateMachine
 
     void RunStateMachine() { ProcessWork(); };
 
+    /// Wait for the supplied amount of time or for interruption.
+    /// If interrupted, returns false, otherwise true.
+    /// @param duration wait duration
+    template<class Rep, class Period>
+    bool WaitFor(std::chrono::duration<Rep, Period> const& duration)
+    {
+        std::unique_lock<std::mutex> lock(fInterruptedMtx);
+        return !fInterruptedCV.wait_for(lock, duration, [&] { return fInterrupted.load(); }); // return true if no interruption happened
```

Comment:
I think you can miss a notification here, e.g. if you wake up spuriously and enter the lambda, read `fInterrupted`, then sleep. Second thread sets `fInterrupted` and wakes you up, but are still in the lambda. You exit the lambda and go to sleep, having consumed the notification at the wrong time. Or?

Reply:
before evaluating the lambda, wait will reacquire the lock, and release it after evaluation. So other threads cannot change fInterrupted (if they lock properly).

Reply:
it's ⚛️ic. :)

Reply:
> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread.

https://en.cppreference.com/w/cpp/thread/condition_variable

Reply:
You're right, missed that point. Adding lock on the writing side.

Reply:
Fixed.